src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,176 @@
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new NullPointerException("The list parameter cannot be null.");
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> setNode = findNode(index);
+        T oldValue = setNode.value;
+        setNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removalNode = findNode(index);
+        if (removalNode.prev == null) {
+            head = removalNode.next;
+        } else {
+            removalNode.prev.next = removalNode.next;
+        }
+        if (removalNode.next == null) {
+            tail = removalNode.prev;
+        } else {
+            removalNode.next.prev = removalNode.prev;
+        }
+        size--;
+        return removalNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (isEqual(currentNode.value, object)) {
+                if (currentNode.prev == null) {
+                    head = currentNode.next;
+                } else {
+                    currentNode.prev.next = currentNode.next;
+                }
+                if (currentNode.next == null) {
+                    tail = currentNode.prev;
+                } else {
+                    currentNode.next.prev = currentNode.prev;
+                }
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isEqual(T value, T obj) {
+        return Objects.equals(obj, value);
```

Comment:
Please write your implementation of null-safe comparison as usage of Objects class is not allowed for this task